pkg/backends/calico/routes.go

```diff
@@ -14,31 +14,265 @@
+}
+
+// onSvcAdd is called when a k8s service is created
+// - creates a serviceEndpoints object and places it in the map
+func (rg *routeGenerator) onSvcAdd(obj interface{}) {
+	log.Info("+onSvcAdd")
+	rg.Lock()
+	defer rg.Unlock()
+
+	svc := obj.(*v1.Service)
+	key := hash(&svc.ObjectMeta)
+	log.WithFields(log.Fields{
+		"key":       key,
+		"serviceIP": svc.Spec.ClusterIP,
+	}).Info("+onSvcAdd")
+	rg.svcEpMap[key] = serviceEndpoints{
+		serviceIP:   svc.Spec.ClusterIP,
+		endpointIPs: make(map[string]bool),
+	}
+
+	// no need to call updateRoutes since the serviceEndpointIPs are empty
+}
+
+// onSvcUpdate is called when a k8s service is updated
+// - clusterIP is an immutable field
+// - only change this method will handle is name/space change
+func (rg *routeGenerator) onSvcUpdate(oldObj, newObj interface{}) {
+	log.Info("+onSvcUpdate")
+	oldSvc, newSvc := oldObj.(*v1.Service), newObj.(*v1.Service)
+
+	// pass if the name/space fields are the same
+	if oldSvc.ObjectMeta.Namespace == newSvc.ObjectMeta.Namespace && oldSvc.ObjectMeta.Name == newSvc.ObjectMeta.Name {
+		return
+	}
+
+	rg.Lock()
+	defer rg.Unlock()
+
+	// update the map
+	oldSvcKey := hash(&oldSvc.ObjectMeta)
+	rg.svcEpMap[hash(&newSvc.ObjectMeta)] = rg.svcEpMap[oldSvcKey]
+	delete(rg.svcEpMap, oldSvcKey)
+
+	// no need to call updateRoutes since no ips actually changes
+}
+
+// onSvcUpdate is called when a k8s service is deleted
+// - removes the serviceEndpoint object from the map
+func (rg *routeGenerator) onSvcDelete(obj interface{}) {
+	log.Info("+onSvcDelete")
+	rg.Lock()
+	defer rg.Unlock()
+
+	svc := obj.(*v1.Service)
+	delete(rg.svcEpMap, hash(&svc.ObjectMeta))
+	rg.updateRoutes()
+}
+
+// waitUntilExists continuously sleeps until the object is found in the svcEpMap
+func (rg *routeGenerator) waitUntilExists(key string) {
+	for {
+		if _, exists := rg.svcEpMap[key]; exists {
+			break
+		}
+		rg.Unlock()
+		time.Sleep(time.Second)
+		rg.Lock()
+		continue
+	}
+}
+
+// onEPAdd is called when a k8s endpoint is created
+// - adds the endpointIPs to the corresponding serviceEndpoints object in the map
+// - sleeps for one second if the serviceEndpoint object does not yet exist
+func (rg *routeGenerator) onEPAdd(obj interface{}) {
+	log.Info("+onEPAdd")
+	ep := obj.(*v1.Endpoints)
+	rg.Lock()
+	defer rg.Unlock()
+
+	// wait until the serviceEndpoint obj exists
+	svcEpKey := hash(&ep.ObjectMeta)
+	rg.waitUntilExists(svcEpKey)
+
+	log.WithFields(log.Fields{
+		"key": svcEpKey,
+	}).Info("onEPAdd")
+
+	// iterate over the subsets and addresses to add the endpoints
+	// that are on this node
+	for _, subset := range ep.Subsets {
+		for _, address := range subset.Addresses {
+			log.WithFields(log.Fields{
+				"addr": address.Hostname,
+			}).Info("onEPAdd")
+			if address.NodeName == nil || *address.NodeName != rg.nodeName {
+				continue
+			}
+			log.Info("onEPAdd: added")
+			rg.svcEpMap[svcEpKey].endpointIPs[address.Hostname] = true
```

Comment:
Wondering here:
- what do we use `address.Hostname` for?  don't we just need a single bool, for whether there is an local active endpoint or not?
- does it matter that https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpointaddress-v1-core says that nodeName is "Optional: Node hosting this endpoint. This can be used to determine endpoints local to a node." ?  If the API says it's optional, can we rely on it being there?

Reply:
@neiljerram yeah, it is optional but I believe in the cases we care about it will be set.

We'll want to think hard about what we should do when it is nil. It's possible to have Endpoints which are backed by things which are not Kubernetes pods. In that case, the node will be nil. There may be others as well, not sure.